File: AdventDay5.cpp
#include <bits/stdc++.h>
using namespace std;

long long partOne() {
    int T;
    if (!(cin >> T)) return 0;

    vector<int> X(T), Y(T);
    for (int i = 0; i < T; ++i) {
        string s;
        cin >> s;
        size_t bar = s.find('|');
        if (bar == string::npos) continue;
        int xVal = stoi(s.substr(0, bar));
        int yVal = stoi(s.substr(bar + 1));
        X[i] = xVal;
        Y[i] = yVal;
    }

    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    long long sumMid = 0;
    string line;
    while (getline(cin, line)) {
        vector<int> pages;
        string token;
        stringstream ss(line);
        while (getline(ss, token, ',')) {
            if (token.empty()) continue;
            pages.push_back(stoi(token));
        }
        if (pages.empty()) continue;

        unordered_map<int, int> pos;
        for (int i = 0; i < (int)pages.size(); ++i) pos[pages[i]] = i;

        bool ok = true;
        for (int i = 0; i < T && ok; ++i)
            if (pos.count(X[i]) && pos.count(Y[i]) && pos[X[i]] >= pos[Y[i]]) ok = false;

        if (ok) sumMid += pages[pages.size() / 2];
    }
    return sumMid;
}

long long partTwo() {
    return partOne();
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << partOne() << '\n';
    return 0;
}

File: Advent_Day_5.py
def day5fileParser():
    try:
        with open("data", "r") as f:
            lines = [line.strip() for line in f if line.strip()]
    except FileNotFoundError:
        return [], []

    if not lines:
        return [], []

    split_idx = 0
    while split_idx < len(lines) and '|' in lines[split_idx]:
        split_idx += 1

    rules = lines[:split_idx]
    updates = lines[split_idx:]
    return rules, updates


def partOne():
    rules, updates = day5fileParser()
    T = len(rules)
    X = [0] * T
    Y = [0] * T

    for i in range(T):
        s = rules[i]
        if '|' not in s:
            continue
        bar = s.find('|')
        X[i] = int(s[:bar])
        Y[i] = int(s[bar + 1:])

    sumMid = 0

    for line in updates:
        vals = line.split(',')
        pages = [int(v) for v in vals if v]
        if not pages:
            continue

        pos = {val: idx for idx, val in enumerate(pages)}

        ok = True
        for i in range(T):
            if X[i] in pos and Y[i] in pos and pos[X[i]] >= pos[Y[i]]:
                ok = False
                break

        if ok:
            sumMid += pages[len(pages) // 2]

    return sumMid


def partTwo():
    rules, updates = day5fileParser()

    return 0


if __name__ == "__main__":
    print(partOne())
    # print(partTwo())
